src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -0,0 +1,292 @@
+        {
+            _session = session;
+            _scope = scope;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void AddAdditionalAssets(CustomAsset asset)
+        {
+            _scope.AddAdditionalAsset(asset, _cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            _scope.Dispose();
+            _session.Dispose();
+        }
+
+        public Task InvokeAsync(string targetName, params object[] arguments) =>
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
+    }
+
+    /// <summary>
+    /// Helper type for common session case
+    /// 
+    /// This will let one to hold onto <see cref="RemoteHostClient.Session"/> for a while.
+    /// this helper will let you not care about remote host being gone while you hold onto the session if that ever happen
+    /// 
+    /// and also make sure state is correct even if multiple threads call TryInvokeAsync at the same time. but this 
+    /// is not optimized to handle highly concurrent usage. if highly concurrent usage is required, either using
+    /// <see cref="RemoteHostClient.Session"/> direclty or using <see cref="SolutionAndSessionHolder"/> would be better choice
+    /// </summary>
+    internal sealed class KeepAliveSessionHolder
+    {
+        private readonly SemaphoreSlim _gate;
+        private readonly IRemoteHostClientService _remoteHostClientService;
+        private readonly CancellationToken _cancellationToken;
+
+        private readonly string _serviceName;
+        private readonly object _callbackTarget;
+
+        private RemoteHostClient _client;
+        private RemoteHostClient.Session _session;
+
+        public KeepAliveSessionHolder(RemoteHostClient client, RemoteHostClient.Session session, string serviceName, object callbackTarget, CancellationToken cancellationToken)
+        {
+            Initialize_NoLock(client, session);
+
+            _gate = new SemaphoreSlim(initialCount: 1);
+            _remoteHostClientService = client.Workspace.Services.GetService<IRemoteHostClientService>();
+            _cancellationToken = cancellationToken;
+
+            _serviceName = serviceName;
+            _callbackTarget = callbackTarget;
+        }
+
+        public void Shutdown()
```

Comment:
This still feels very bizarre.  It's like having IList, and then having a bunch of different list impl that don't derive from it.

It seems like you have a simple to understand concept: the session.  It allows you to send remote requests to the OOP server.

Then, you have helper functions that give you sessions with varying levels of cost/complexity:

1. A basic session that can just remote commands over, but doesn't really do any state synchronization.  Caller gets it, sends commands, and disposes when done.
2. A state synchronizing session that allows the client/server to agree on a Solution snapshot.  Caller gets it, sends commands, and disposes when done.
3. A session that is intended to be kept around for a long time.  Caller gets it, sends commands, and eventually disposes when it no longer needs it.

In all events, the shape here is the same.  The individual details are different, but that is handled by how you ask for the instance to be created/acquired.  The caller doesn't need to then interact with different types.  They all just work with a basic ISession (or Session) type that only exposes hte behavior that is required.